GitUI/CommandsDialogs/FormBrowse.cs

```diff
@@ -952,9 +953,18 @@ private void InternalInitialize()
 
                     _formBrowseMenus.InsertRevisionGridMainMenuItems(repositoryToolStripMenuItem);
 
-                    // Request all branches if left panel is shown
-                    IDictionary<string, AheadBehindData> aheadBehindData = _aheadBehindDataProvider?.GetData(MainSplitContainer.Panel1Collapsed ? RevisionGrid.CurrentBranch.Value : "");
-                    toolStripButtonPush.DisplayAheadBehindInformation(aheadBehindData, RevisionGrid.CurrentBranch.Value);
+                    if (AppSettings.ShowAheadBehindData)
+                    {
+                        bool isAheadBehindDisplayedForAllBranches = !MainSplitContainer.Panel1Collapsed && (AppSettings.RepoObjectsTreeShowRemotes || AppSettings.RepoObjectsTreeShowBranches);
+                        string currentBranch = RevisionGrid.CurrentBranch.Value;
+                        ThreadHelper.FileAndForget(async () =>
+                        {
+                            string aheadBehindBranch = isAheadBehindDisplayedForAllBranches ? "" : currentBranch;
+                            IDictionary<string, AheadBehindData> aheadBehindData = _aheadBehindDataProvider?.GetData(aheadBehindBranch);
+                            await this.SwitchToMainThreadAsync();
+                            toolStripButtonPush.DisplayAheadBehindInformation(aheadBehindData, currentBranch);
```

Comment:
> For me ahead-behind takes about 500ms.
> I found that annoy, pushing a fixup.

If it is the case, please really consider doing the call here only for the current branch.
I know it will be an extra git call but it has the advantage to simplify the code (here and the new code you wish to add) and the call is pretty quick (60 to 90ms) so you have the benefit that the information displayed in the push button is always accurate (and it's a lot better when you open a big repository than having to wait more than 6s).
And maybe in this case, we could even remove the newly added `ThreadHelper.FileAndForget()` simplifying even more the code...

```suggestion
                        string currentBranch = RevisionGrid.CurrentBranch.Value;
                        ThreadHelper.FileAndForget(async () =>
                        {
                            IDictionary<string, AheadBehindData> aheadBehindData = _aheadBehindDataProvider?.GetData(currentBranch);
                            await this.SwitchToMainThreadAsync();
                            toolStripButtonPush.DisplayAheadBehindInformation(aheadBehindData, currentBranch);
```

Reply:
When testing this solution, I noticed there is a race condition:
When there is not a lot of local branches, the ahead-behind call for all branches arrives first and so execute quickly. The consequence is that the call specific for current branch is taking result from the cache and no 2nd git command is run.
When there are a lot of local branches (2k), the call for push button arrives first and so the lightest command is run first.

After testing with repositories with different branches count, it seems that the switch seems to happen when you have between 2k to 5k branches (where ahead-behind call takes more than 1.5s).
So, in the light of this information, I think we must:
* keep the `ThreadHelper.FileAndForget()` and your button size workaround contrary to what I just said earlier
* always try to get ahead-behind **for the current branch** here. If there are really a lot of branches, you will get the needed information quicker (because for me the ahead-behind in the push button is much more important than the one in the tree view). It's a kind of clue than the command will be long to run. In the other case, the command for the current branch won't be run and will take data from cache.

So I think you should accept my code suggestion and we are good to go.

Reply:
might be a usecase for `ExclusiveTaskRunner`

Reply:
I don't see a way how `ExclusiveTaskRunner` could improve the situation here. I think I understand what is the purpose but I don't see added value.

Except if we integrate all the retrieve thread logic into the provider with always launching thread to retrieve current branch before the one to retrieve all branches data. And we can start the threads earlier (just after retrieving data for revision grid) and accessing the data will wait completion of the thread(s).

A big refactoring but I don't think we will gain some time because the commands are launched pretty quickly in fact...

I think we are good enough now.

Reply:
The total time from the reset to the Git command is done is about 500 ms (so not just the Git command) - if the clear old data change is not reverted.
So current changes should be OK.